src/createRoom.ts

```diff
@@ -221,6 +221,14 @@ export default async function createRoom(opts: IOpts): Promise<string | null> {
     let roomId;
     return client.createRoom(createOpts).finally(function() {
         if (modal) modal.close();
+    }).catch(function(err) {
+        if (err.httpStatus === 403 && err.errcode == "M_UNKNOWN") {
```

Comment:
this feels like a very broad catch, maybe a better solution would be on the Synapse side to ignore `visibility` when that specific homeserver.yaml option is specified?

Reply:
That's a good point, but I worry that altering server behaviour may cause trouble for clients.

Since the spec doesn't say otherwise, it seems that if a `/createRoom` request succeeds, then the requesting client should have a guarantee that all settings they wanted the new room to have were applied. Otherwise, clients may make false assumptions about the settings of the room they just created.

An example of when this could be a problem is for appservices. Say an appservice user wants to create a published room on a Synapse homeserver that, unbeknownst to the appservice, denies them room publishing privileges. If the appservice user calls `/createRoom` with `"visibility": "public"`, and Synapse creates the room but doesn't publish it, all the appservice sees is a successful `200` response (and the new room's ID in the payload), so it doesn't know whether the created room actually has public visibility. This could be bad if the room the appservice wanted to publish would be useless as a private room (like an alias-free room that the appservice doesn't invite anyone else to).

This wouldn't be as much of a problem if Synapse (or the CS spec) had a server API for requesting what a user's publish privileges are, but AFAICT there isn't one.

The alternative is for Synapse to respond with a dictionary of what settings were changed from the original request, but that won't help a user (sapient or otherwise) that wouldn't have wanted the room to be created at all if its requested settings couldn't be used.

So the current situation of Synapse giving a failed response seems like the best approach, since clients (sapient or otherwise) can choose to respond to that failure however they like.

Reply:
Then there being a more concrete error code for this would be better, because the 403/M_UNKNOWN could be transient and retrying with publishing might have actually worked.

Reply:
The issue is that there are a few potential reasons why Synapse may respond to `/createRoom` with 403; there's even a comment about it: https://github.com/matrix-org/synapse/blob/428174f90249ec50f977b5ef5c5cf9f92599ee0a/synapse/handlers/room.py#L771

It would be useful if this case could use `M_FORBIDDEN`, but the other 403 responses feel as though they could also use it too, and if they were to use it, clients wouldn't be able to distinguish them & we'd be right back where we started.

Maybe this isn't stable or spec-compliant, but my latest commit uses the error message field to precisely check what the error was.